cpp/src/arrow/compute/kernels/base_arithmetic_internal.h

```diff
@@ -485,6 +485,22 @@ struct NegateChecked {
   }
 };
 
+struct Exp {
+  template <typename T, typename Arg>
+  static constexpr enable_if_floating_value<Arg, T> Call(KernelContext*, Arg exp,
+                                                         Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    return std::exp(exp);
+  }
+
+  template <typename T, typename Arg>
+  static constexpr enable_if_decimal_value<Arg, T> Call(KernelContext*, Arg exp,
+                                                        Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    return exp.Exp();
```

Comment:
Does this actually exist? I don't see a definition for it in (Basic)Decimal128

Reply:
I guess it's tested below so it should clearly exist…

Reply:
Ah, no, this isn't being used at all - decimals are being converted to floats first. Can we remove this?

Reply:
QQ: David, are you asking if the `enable_if_decimal_value` exists? I'm not sure I understand your comment, sorry...

Reply:
@drabastomek {{Decimal128::Exp}} doesn't exist, so this definition wouldn't compile.

Reply:
Ah... I see. Ok, but the Substrait still can call Acero with a decimal value. I understand it'd be converted to float first to do the `Exp` but would we rather error on this or keep it? I'm happy to remove this and the relevant test just thinking out loud.

Reply:
The code here is dead code. We should remove it.

It doesn't matter what Substrait does; we currently implement most of these kernels by casting to float first. If that's not OK for Substrait, then we'll have to come back to implement actual decimal versions of these kernels.

Reply:
FWIW, we should not be allowing this kind of implicit cast and should instead reject a plan containing a decimal exp call.  If the producer wants to allow the cast to float they can allow it.

However, this issue is beyond this PR.  I've created https://issues.apache.org/jira/browse/ARROW-18193